modules/node_modules/@frogpond/ccc-bonapp/index.js

```diff
@@ -1,29 +1,275 @@
+	virtualConsole.on('jsdomError', (err) => {
+		let messagesToSkip = [
+			'Uncaught [ReferenceError: wp is not defined]',
+			'Uncaught [ReferenceError: jQuery is not defined]',
+		]
+		if (messagesToSkip.includes(err.message)) {
+			return
+		}
+		console.error(err)
+	})
+	return JSDOM.fromURL(url, {
+		runScripts: 'dangerously',
+		virtualConsole,
+	})
+}
+
+const zodYesNo = z.union([z.literal('Y'), z.literal('N')])
+const zodNumericString = z.string().regex(/d*/)
+const zodCurrencyString = z.string() //.regex(/^\$.*/)
+const zodHtmlString = z.string()
+const zod24Time = z.string().regex(/\d\d:\d\d/)
+const zodNumericBoolean = z.number().gte(0).lte(1)
+
+const MonotonyContainer = z.object({
+	id: z.string(),
+	name: z.string(),
+	short_name: z.optional(z.null()),
+	image: z.string(),
+})
+
+const NutritionContainer = z.object({
+	kcal: zodNumericString,
+	well_being: z.string(),
+	well_being_image: z.string(),
+})
+
+const NutritionDetailType = z.object({
+	label: z.string(),
+	value: z.string(),
+	unit: z.string(),
+})
+
+const NutritionDetailContainer = z.object({
+	calories: z.optional(NutritionDetailType),
+	servingSize: z.optional(NutritionDetailType),
+	fatContent: z.optional(NutritionDetailType),
+	saturatedFatContent: z.optional(NutritionDetailType),
+	transFatContent: z.optional(NutritionDetailType),
+	cholesterolContent: z.optional(NutritionDetailType),
+	sodiumContent: z.optional(NutritionDetailType),
+	carbohydrateContent: z.optional(NutritionDetailType),
+	fiberContent: z.optional(NutritionDetailType),
+	sugarContent: z.optional(NutritionDetailType),
+	proteinContent: z.optional(NutritionDetailType),
+})
+
+const CafeMenuItem = z.object({
+	connector: z.string(),
+	cor_icon: z.union([z.array(z.unknown()), z.record(z.string())]),
+	description: z.string(),
+	id: z.unknown(),
+	label: z.string(),
+	monotony: z.union([z.object({}), MonotonyContainer]),
+	nutrition: NutritionContainer,
+	nutrition_details: z.optional(NutritionDetailContainer),
+	nutrition_link: z.string(),
+	options: z.union([z.array(z.unknown()), z.record(z.unknown())]),
+	price: zodCurrencyString,
+	rating: zodNumericString,
+	special: zodNumericBoolean,
+	station: zodHtmlString,
+	sub_station: z.string(),
+	sub_station_id: zodNumericString,
+	sub_station_order: zodNumericString,
+	zero_entree: zodNumericString,
+})
+
+const BamcoStation = z.object({
+	order_id: z.string(), // sort on order_id instead of sorting on id
+	id: zodNumericString,
+	label: z.string(),
+	price: zodCurrencyString,
+	note: z.string(),
+	soup: z.unknown(),
+	items: z.array(z.string()),
+})
+
+const BamcoDayPart = z.object({
+	abbreviation: z.string(),
+	endtime: zod24Time,
+	endtime_formatted: z.string(),
+	id: z.unknown(),
+	label: z.string(),
+	message: z.union([z.literal(''), z.string()]),
+	starttime: zod24Time,
+	starttime_formatted: z.string(),
+	time_formatted: z.string(),
+	stations: z.array(BamcoStation),
+})
+
+const CafeInfo = z.object({
+	name: z.string(),
```

Comment:
I like that we're using zod in this PR! Would it make sense to treat more of our types as optional because we don't own the contract? What are your thoughts on making most things optional variants and falling back to empty defaults when we return data?

Reply:
Current status:

~|~
--|--
![Simulator Screenshot - iPhone 15 Pro - 2024-05-03 at 23 27 51](https://github.com/frog-pond/ccc-server/assets/5240843/1797c314-2065-452f-89df-7498036e1d02) | ![Simulator Screenshot - iPhone 15 Pro - 2024-05-03 at 23 21 47](https://github.com/frog-pond/ccc-server/assets/5240843/02365e44-0d43-40a1-b078-61214da6ab50)
![Simulator Screenshot - iPhone 15 Pro - 2024-05-03 at 23 21 49](https://github.com/frog-pond/ccc-server/assets/5240843/6fba149c-d88d-471e-8cfd-45efe0c13bcd) | ![Simulator Screenshot - iPhone 15 Pro - 2024-05-03 at 23 29 03](https://github.com/frog-pond/ccc-server/assets/5240843/b6fd872e-87b9-4822-9f25-5ca66938fd2e)

`xxx GET /v1/food/named/menu/the-cage 500`
```
  ZodError: [
    {
      "code": "invalid_type",
      "expected": "object",
      "received": "array",
      "path": [
        "cor_icons"
      ],
      "message": "Expected object, received array"
    }
  ]
  
at get error [as error] (zod/lib/index.mjs:587:31)
at ZodObject.parse (zod/lib/index.mjs:692:22)
at menu (@frogpond/ccc-bonapp/index.js:260:32)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async cageMenu (@frogpond/ccci-stolaf-college/v1/menu/index.js:84:13)
at async bodyPkoa-bodyparser/index.js:78:5)
at asynckoa-etag/index.js:27:5)
atkoa-conditional-get/index.js:15:5
at async compressMiddlkoa-compress/lib/index.js:56:5)
at async lkoa-logger/index.js:67:7)
```

`GET /v1/food/named/menu/the-pause 204`
```
A problem occured while loading: 
Error: ["cafe-menu","hosted","food/named/menu/the-pause"]
data is undefined
```

`GET /v1/food/named/cafe/weitz 500`
```
  ZodError: [
    {
      "code": "invalid_type",
      "expected": "object",
      "received": "undefined",
      "path": [],
      "message": "Required"
    }
  ]
at get error [as error] (zod/lib/index.mjs:587:31)
at ZodObject.parse (zod/lib/index.mjs:692:22)
at cafe (@frogpond/ccc-bonapp/index.js:178:32)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async weitzCafe (@frogpond/ccci-stolaf-college/v1/menu/index.js:138:13)
at async bodyParser (koa-bodyparser/index.js:78:5)
at async etag (koa-etag/index.js:27:5)
at async koa-conditional-get/index.js:15:5
at async compressMiddleware (koa-compress/lib/index.js:56:5)
at async logger (koa-logger/index.js:67:7)
```

Reply:
@drewvolz We can do that for the "input" types, sure! We should also hook up CCC to Sentry, so we can be notified if the upstream schema changes on us.

Reply:
@hawkrives I can handle hooking us up to Sentry — it has been on my list for a bit. I'll do that on a separate branch.